### DISCUSSION TITLE
Rustdesk keyboard auto-capture and lock (hook for wayland)

### DISCUSSION THREAD

**Discussion by ghost** [rustdesk/rustdesk#5117] [Feature Request]
### Bug Description

When in Linux (Debian 11 - Wayland), remotely connecting to a client (Windows or Linux) and pressing the **Win** key (which pulls up the Start Menu in Windows for example) results into showing all open applications in my Linux host.

Is it possible to implement or modify manually Rustdesk on my own to capture the keyboard like Virtualbox & QEMU does and keep me inside the session window?

It would be perfect to use the same mechanism and lock the session window, so we have to press a release combination like **L-Ctrl + L-Alt**, to be able to interact with our host again.

### How to Reproduce

Connect to remote client using a Debian 11 host under Wayland and Windows 11 as a client, and press the **Win** key. Instead of pulling up the client's Start Menu, it'll interact the the host.

### Expected Behavior

Pressing any key should only interact with the client session and not the host.

### Operating system(s) on local side and remote side

Debian 11 -> Windows 11

### RustDesk Version(s) on local side and remote side

1.2.1 -> 1.2.1

### Screenshots

Screenshots not applicable for specific problem, self explanatory.

**Comment by rustdesk** [comment]
Are you sure virtualbox in Wayland can capture system keys?

**Comment by ghost** [comment]
Very older versions couldn't, and I thought the same but the last few versions work. Debian 11 Wayland with Virtualbox v7.0 works correctly, I can press the Win key to open the Start Menu in the Windows 10/11 client and any other combination. Only when hitting L-Ctrl + L-Alt releases the lock and I can continue with the host.

**Reply by rustdesk** [reply]
Got it, Thanks.

**Reply by rustdesk** [reply]
@fufesou let's put in our todo list. This should have higher priroity than the other controlled side wayland issues.

**Comment by ghost** [comment]
> @fufesou let's put in our todo list.

Thank you guys, appreciate and love the project!

**Comment by SergeyMy** [comment]
@engineer22 
are you sure that the button "super" is not intercepted? (important the **mouse pointer must be in the focus of the window rustdesk**)
in X11 everything is fine

**Comment by bayazidbh** [comment]
Would this portal help: flatpak/xdg-desktop-portal/pull/714

**Comment by thinkORo** [comment]
@rustdesk Is there any news on this yet? I am currently checking whether Rustdesk could replace “RDP+Remmina”, but the auto-capture feature is a bit annoying, as I now have to use the mouse very often to get back to the host from the remote environment or to a different remote machine. Is there any workaround?

**Comment by freddyli** [comment]
Am I understanding this correctly, that with RustDesk client running on Wayland, it is not possible to capture the Windows key or ALT-TAB properly to forward them only to the remote host? :thinking:

**Comment by thinkORo** [comment]
Hi @freddyli : it is the other way around. The keyboard is auto-captured and there is no hotkey (or at least: I couldn't find any) to release the keyboard from the remote connection.

In my case: The remote connection is on virtual desk 2 (on my local machine) and when I press the hotkey to switch back to virtual desk 1 the hot key is captured in the remote only. And I have to use my mouse to go back to virtual desk 1 from 2.

Hope the explanation makes sense (to you)?!

**Reply by msilveirabr** [reply]
It looks like you have a very peculiar use of rustdesk... are  you remoting to localhost?

**Reply by thinkORo** [reply]
No, I only want to explain that the keyboard capture cannot escape from the remote connection. But I think it was explained to complicated. I opened a remote connection from my local Linux machine to my remote PC (with any supported OS). And I cannot press any keyboard combination to control my local Linux desktop. That it. Sorry for confusing.

**Reply by msilveirabr** [reply]
Now that made sense 🙂
It looks pretty much like my setup... but Fedora42+Gnome
I use 3 monitors, 2 + virtual desktops... what I do to "ungrab" the keyboard from Rustdesk on Xorg is to move the mouse to the very TOP center of the screen and press ALT+TAB, or move the mouse to another screen/app, off from rustdesk, then press ALT+TAB, WindowsKey, whatever...

Are you sure you get keyboard grabbed in Wayland? Try running xeyes, open rustdesk ( no need to connect to remote ), then move the mouse over rustdesk window, if the eyes from xeyes move, that means you're either using Xorg ( not Wayland )  OR rustdesk is running in X11 mode instead of Xwayland. 🤷‍♂️

**Reply by thinkORo** [reply]
To move the mouse outside of Rustdesks remote window is not the problem (because it's working as you described). But also not the solution because I don't want to use the mouse for this purpose.

I think I still have to wait for an implementation like the one of VirtualBox where one can define an escape key.

**Reply by msilveirabr** [reply]
Right, I understand. 
Anyway... I have to ask again: Are you sure the keyboard grab is working on Xwayland??? (GNOME?)
If the answer is yes, then I wonder what you've done to achieve this... or what are the differences between Fedora and Ubuntu implementation...

**Reply by thinkORo** [reply]
I've tested xeyes as you mentioned. And it seems that I am not using Wayland (on my Arch Linux machine). And I read the original post from ghost again. And his/her problem is the other way around as mine. I want to escape and he/her want to be caught.

**Comment by msilveirabr** [comment]
Any news on this front? The lack of keyboard capture on wayland really happens ( Fedora42 here, If I login using Xorg, capture is OK )...

**Reply by msilveirabr** [reply]
BTW, Remmina 1.4.40 is able to behave in Wayland exactly like rustdesk behaves in X11/Xorg.
With the extra of providing the feature @thinkORo asked for, by using `<Super>+<ESC>`  to release the keyboard, and grabbing automatically when clicking in the remote screen again :)

This behavior would be awesome in @rustdesk !

**Reply by msilveirabr** [reply]
Here's an old, but interesting discussion about remmina with mutter: https://gitlab.gnome.org/GNOME/mutter/-/issues/2450

**Comment by AntoineGS** [comment]
I am also running this problem as I use a window manager on my remote (Windows) environment. Using the Win/Super key is not relayed and prevents me from using all those shortcuts :(
Edit: While flagged as an unstable Wayland API, this might be the way to do it: https://wayland.app/protocols/keyboard-shortcuts-inhibit-unstable-v1

**Comment by AntoineGS** [comment]
To anyone using Hyperland, I was able to create a script that automatically disables the keymaps for the local machine whenever a rustdesk window is active. You can find both files [here](https://github.com/AntoineGS/configurations/commit/70c51930cba023dce997e8a887f87ab8ab81bc8f).
The following steps should work:
- Change the .service file to ensure the correct path to the .sh is set
- Make sure to run `sudo chmod +x /path/to/sh/file` 
- Place the .service file in ~/.config/systemd/user
- Run `systemctl --user daemon-reload` and `systemctl --user enable --now watch-rustdesk-submap.service`
- You can check if it loaded successfully using `journalctl --user -u watch-rustdesk-submap.service -f`

**Reply by NutchapolSal** [reply]
Alternatively for KDE Plasma, you can set "Ignore global shortcuts" for Rustdesk windows using System Settings > Window Management > Window Rules.
Add a new rule, use "Detect Window Properties", pick a Rustdesk window, then add "Whole window class".
Close the dialog, then click on "Add property...", search for "Ignore global shortcuts", add it, set it to Yes, then apply.

<img width="1433" height="742" alt="Screenshot_20260608_101246" src="https://github.com/user-attachments/assets/c7d7359d-1e92-4a7b-96fa-a8c02b000088" />


Focusing on Rustdesk should now no longer trigger host's Alt+Tab and Win key

**Comment by dan-passaro** [comment]
I've also run into this problem. It makes it so that I can't use Super+Tab to change applications in macOS (among other things) from a Linux host using Wayland. For now I work around the issue by using Xorg instead, but it seems like the Linux community is trying to sunset Xorg so having this functionality for Wayland would be great.

**Comment by fufesou** [comment]
https://github.com/rustdesk/rustdesk/pull/14302
